# Diabetic Hospital Readmission Prediction
## Notebook 01 — ETL Pipeline & Data Cleaning

### What This Notebook Does:
- Loads 101,766 raw patient records from 130 US hospitals
- Removes duplicate patient encounters — keeps first visit only
- Removes deceased and hospice patients from the dataset
- Preserves A1C blood test signal as 3 binary features
- Cleans missing values and removes unusable columns
- Creates binary target variable readmitted_30
- Converts age ranges to numeric midpoints
- Simplifies 700+ ICD-9 diagnosis codes into 9 disease categories
- Applies One-Hot Encoding for diagnosis columns
- Applies Label Encoding for binary and ordinal columns
- Keeps medical_specialty as text for Target Encoding in modeling
- Joins CDC PLACES SDOH data at state level
- Saves clean merged dataset for feature engineering

### Datasets Used:
- UCI Diabetes 130-US Hospitals (1999-2008) — 101,766 rows
- CDC PLACES County Health Data — 229,298 rows

### Output Files:
- data/processed/diabetic_cleaned.csv — 69,973 unique patients
- data/processed/diabetic_with_sdoh.csv — with 6 SDOH features
- data/processed/sdoh_state_averages.csv — state level SDOH data

In [52]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder
import warnings
warnings.filterwarnings('ignore')

print("Libraries loaded!")

Libraries loaded!


We import all libraries needed for this notebook. pandas handles data loading and manipulation. numpy handles numerical operations. matplotlib and seaborn are for any visualizations. LabelEncoder converts text categories to numbers. warnings.filterwarnings('ignore') suppresses non-critical warnings to keep output clean.

In [53]:
# Loading the main dataset
df = pd.read_csv("../data/raw/diabetic_data.csv")

print(f'Rows: {len(df):,}')
print(f'Columns: {df.shape[1]}')
print(f'\nFirst 3 rows:')
df.head(3)

Rows: 101,766
Columns: 50

First 3 rows:


,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,2278392,8222157,Caucasian,Female,[0-10),?,6,25,1,1,...,No,No,No,No,No,No,No,No,No,NO
1,149190,55629189,Caucasian,Female,[10-20),?,1,1,7,3,...,No,Up,No,No,No,No,No,Ch,Yes,>30
2,64410,86047875,AfricanAmerican,Female,[20-30),?,1,1,7,2,...,No,No,No,No,No,No,No,No,Yes,NO


We load the raw UCI Diabetes 130-US Hospitals dataset. This is the starting point of the entire project. The dataset contains 101,766 patient encounters collected from 130 US hospitals between 1999 and 2008. Each row is one hospital visit — not one unique patient. The dataset has 50 columns covering patient demographics, diagnoses, medications, lab results, and the readmission outcome. We print the shape first to confirm the file loaded correctly before doing anything else.

#### Dataset Column Descriptions

##### Patient Identifiers
- encounter_id: Unique ID for each hospital visit
- patient_nbr: Unique ID for each patient
##### Patient Demographics
- race: Patient race — Caucasian, AfricanAmerican, Hispanic, Asian, Other
- gender: Patient gender — Male, Female, Unknown
- age: Age in 10-year brackets — [0-10), [10-20) ... [90-100)
##### Admission Information
- admission_type_id: Type of admission — 1=Emergency, 2=Urgent, 3=Elective
- discharge_disposition_id: Where patient went after discharge — 1=Home, 11=Expired
- admission_source_id: Where patient came from — 7=Emergency Room, 1=Physician Referral
- time_in_hospital: Number of days stayed in hospital (1 to 14)
- payer_code: Insurance type
##### Clinical Information
- medical_specialty: Admitting doctor specialty — Cardiology, Surgery, etc.
- num_lab_procedures: Number of lab tests performed during visit
- num_procedures: Number of non-lab procedures performed
- num_medications: Number of distinct medications given
- number_outpatient: Outpatient visits in the year before this admission
- number_emergency: Emergency visits in the year before this admission
- number_inpatient: Inpatient visits in the year before this admission
- number_diagnoses: Total number of diagnoses recorded for this visit
##### Diagnosis Codes
- diag_1: Primary diagnosis code (ICD-9 format)
- diag_2: Secondary diagnosis code
- diag_3: Tertiary diagnosis code
##### Lab Results
- max_glu_serum: Maximum glucose serum test result
- A1Cresult: HbA1c blood test result — Norm, >7, >8, or ? (not measured)
##### Diabetes Medications
Each column shows if the drug was prescribed and dosage change:
No = not prescribed, Steady = no change, Up = increased, Down = decreased
- metformin, repaglinide, nateglinide, chlorpropamide, glimepiride
- acetohexamide, glipizide, glyburide, tolbutamide, pioglitazone
- rosiglitazone, acarbose, miglitol, troglitazone, tolazamide
- examide, citoglipton, insulin, glyburide-metformin
- glipizide-metformin, glimepiride-pioglitazone
- metformin-rosiglitazone, metformin-pioglitazone
##### Medication Change Indicators
- change: Was any diabetes medication changed? Ch=Yes, No=No
- diabetesMed: Was any diabetes medication prescribed? Yes or No
##### Target Variable
- readmitted: Was patient readmitted? <30=within 30 days, >30=after 30 days, NO=never

In [54]:
# ============================================================
# Deduplicate Patients — Keep First Encounter Only
# ============================================================
# Problem: Same patient appears multiple times in dataset
# Example: John visited hospital 3 times = 3 rows for John
# When we split randomly:
#   John's visit 1 goes to training set
#   John's visit 3 goes to test set
# Model already learned John's patterns = data leakage!
# Solution: Keep only the FIRST encounter per patient
# ============================================================

print(f'Before deduplication:')
print(f'  Total rows: {len(df):,}')
print(f'  Unique patients: {df["patient_nbr"].nunique():,}')
print(f'  Duplicate encounters: {len(df) - df["patient_nbr"].nunique():,}')

# Sort by encounter_id so earliest visit comes first
df = df.sort_values('encounter_id')

# Keep only first encounter per patient
df = df.groupby('patient_nbr').first().reset_index()

print(f'\nAfter deduplication:')
print(f'  Total rows: {len(df):,}')
print(f'  Each row = one unique patient')
print(f'  Removed: {101766 - len(df):,} duplicate encounters')

Before deduplication:
  Total rows: 101,766
  Unique patients: 71,518
  Duplicate encounters: 30,248

After deduplication:
  Total rows: 71,518
  Each row = one unique patient
  Removed: 30,248 duplicate encounters


In [55]:
# Check missing values (? is used as missing in this dataset)
df_check = df.replace("?", np.nan)

missing = df_check.isnull().sum()
missing_pct = (missing / len(df) * 100).round(2)

missing_df = pd.DataFrame({
    'Missing Count': missing,
    'Missing %': missing_pct
}).sort_values('Missing %', ascending=False)

print("Columns with missing values:")
print(missing_df[missing_df['Missing Count'] > 0])

Columns with missing values:
                   Missing Count  Missing %
weight                     68665      96.01
max_glu_serum              67906      94.95
A1Cresult                  56130      78.48
medical_specialty          34477      48.21
payer_code                 31043      43.41
race                        1948       2.72
diag_3                      1225       1.71
diag_2                       294       0.41
diag_1                        11       0.02


In [56]:
# ============================================================
# Preserve A1C Signal as Binary Flags
# ============================================================
# A1Cresult column has 78% missing values
# But NOT measured is itself clinically meaningful!
#
# What is A1C?
# HbA1c measures average blood sugar over past 3 months
# For diabetic patients it shows long-term glucose control
# Values: Norm=normal, >7=elevated, >8=highly elevated
#
# Why missing matters:
# If A1C was NOT measured the hospital did not assess
# long-term diabetes control — associated with higher
# readmission risk
#
# Solution: Create 3 binary flags BEFORE replacing ?
# A1c_measured = 1 if A1C was tested, 0 if not tested
# A1c_normal   = 1 if result was normal
# A1c_elevated = 1 if result was >7 or >8
# ============================================================

# Check raw values BEFORE replacing ? with NaN
print('Raw A1Cresult values:')
print(df['A1Cresult'].value_counts())

# Create A1C binary flags from raw text values BEFORE replacing ?
df['A1c_measured'] = (df['A1Cresult'] != '?').astype(int)
df['A1c_normal']   = (df['A1Cresult'] == 'Norm').astype(int)
df['A1c_elevated'] = (df['A1Cresult'].isin(['>7', '>8'])).astype(int)

print('A1C binary flags created:')
print(f'  A1c_measured: {df["A1c_measured"].sum():,} ({df["A1c_measured"].mean()*100:.1f}% had A1C measured)')
print(f'  A1c_normal:   {df["A1c_normal"].sum():,} patients had normal A1C')
print(f'  A1c_elevated: {df["A1c_elevated"].sum():,} patients had elevated A1C')
print(f'\nClinical signal preserved with zero missing value problems!')

Raw A1Cresult values:
A1Cresult
>8      7304
Norm    4600
>7      3484
Name: count, dtype: int64
A1C binary flags created:
  A1c_measured: 71,518 (100.0% had A1C measured)
  A1c_normal:   4,600 patients had normal A1C
  A1c_elevated: 10,788 patients had elevated A1C

Clinical signal preserved with zero missing value problems!


In [57]:
## Cleaning the Dataset
# Replace all ? with NaN
df = df.replace('?', np.nan)

# Drop columns with too much missing data
df = df.drop(columns=['weight', 'max_glu_serum', 'A1Cresult', 'payer_code'])

# Fill remaining missing values
df['medical_specialty'] = df['medical_specialty'].fillna('Unknown')
df['race']   = df['race'].fillna(df['race'].mode()[0])
df['diag_1'] = df['diag_1'].fillna('Unknown')
df['diag_2'] = df['diag_2'].fillna('Unknown')
df['diag_3'] = df['diag_3'].fillna('Unknown')

print(f'Dropped columns: weight, max_glu_serum, A1Cresult, payer_code')
print(f'Filled: medical_specialty, race, diag_1, diag_2, diag_3')
print(f'Missing values remaining: {df.isnull().sum().sum()}')
print(f'Dataset shape: {df.shape}')

Dropped columns: weight, max_glu_serum, A1Cresult, payer_code
Filled: medical_specialty, race, diag_1, diag_2, diag_3
Missing values remaining: 0
Dataset shape: (71518, 49)


In [58]:
# ============================================================
# FIX 2: Remove Ineligible Patients
# ============================================================
# Patients who died or went to hospice are coded
# as "not readmitted" because they CANNOT be readmitted.
# They are very different from healthy discharged patients.
# Mixing them into the negative class confuses the model.
#
# Discharge codes that mean patient cannot be readmitted:
# 11 = Expired in hospital
# 13 = Hospice / Medical Facility
# 14 = Hospice / Home
# 19 = Expired at home
# 20 = Expired in Medical Facility
# 21 = Expired — place unknown
# ============================================================

print(f"Before removing ineligible patients: {len(df):,} rows")

ineligible_dispositions = [11, 13, 14, 19, 20, 21]
removed = df[df['discharge_disposition_id'].isin(ineligible_dispositions)]
print(f"Dead/hospice patients found: {len(removed):,}")

df = df[~df['discharge_disposition_id'].isin(ineligible_dispositions)]

print(f"After removing ineligible patients: {len(df):,} rows")
print(f"Removed: {len(removed):,} dead/hospice patients")
print(f"Negative class is now cleaner and model more valid")

Before removing ineligible patients: 71,518 rows
Dead/hospice patients found: 1,545
After removing ineligible patients: 69,973 rows
Removed: 1,545 dead/hospice patients
Negative class is now cleaner and model more valid


When we create our target variable we label all patients not readmitted within 30 days as 0. But some of these patients did not come back because they died or entered hospice — not because they recovered. A model cannot distinguish between a healthy discharge labeled 0 and a death labeled 0. Removing these 1,545 patients makes the negative class clinically meaningful — every patient labeled 0 is someone who was discharged alive and had the opportunity to return but did not.

In [59]:
# Create binary target variable
# 1 = readmitted within 30 days, 0 = not readmitted
df['readmitted_30'] = (df['readmitted'] == '<30').astype(int)

# Drop original readmitted column
df = df.drop(columns=['readmitted'])

# Check target distribution
counts = df['readmitted_30'].value_counts()
pct = df['readmitted_30'].value_counts(normalize=True) * 100

print("Target Variable Distribution:")
print(f"Not Readmitted (0): {counts[0]:,} ({pct[0]:.1f}%)")
print(f"Readmitted <30 (1): {counts[1]:,} ({pct[1]:.1f}%)")
print(f'\nClass imbalance ratio: {counts[0]/counts[1]:.1f}:1')
print(f'\nWhy ROC-AUC is our primary metric:')
print(f'  A model predicting NOT readmitted for everyone')
print(f'  would be {pct[0]:.1f}% accurate but clinically useless')
print(f'  ROC-AUC measures true separation ability')

Target Variable Distribution:
Not Readmitted (0): 63,696 (91.0%)
Readmitted <30 (1): 6,277 (9.0%)

Class imbalance ratio: 10.1:1

Why ROC-AUC is our primary metric:
  A model predicting NOT readmitted for everyone
  would be 91.0% accurate but clinically useless
  ROC-AUC measures true separation ability


The original readmitted column has three values — less than 30 days, more than 30 days, and NO. We convert this to binary because our clinical goal is specifically to identify patients at risk of early readmission within 30 days. Patients readmitted after 30 days are treated the same as never readmitted for this task. The resulting 91% to 9% class imbalance means accuracy is a misleading metric. A model that always predicts not readmitted would be 91% accurate but would miss every single high risk patient. ROC-AUC measures how well the model separates the two classes regardless of imbalance making it the correct metric for this problem.

In [60]:
# Drop ID columns and zero-variance columns
df = df.drop(columns=['encounter_id', 'patient_nbr', 'examide', 'citoglipton'])

print(f'Dropped 4 columns:')
print(f'  encounter_id — unique ID per visit, no predictive value')
print(f'  patient_nbr — unique ID per patient, no predictive value')
print(f'  examide — only 1 unique value (No) across all patients')
print(f'  citoglipton — only 1 unique value (No) across all patients')
print(f'\nDataset shape: {df.shape}')

Dropped 4 columns:
  encounter_id — unique ID per visit, no predictive value
  patient_nbr — unique ID per patient, no predictive value
  examide — only 1 unique value (No) across all patients
  citoglipton — only 1 unique value (No) across all patients

Dataset shape: (69973, 45)


We drop 4 columns that add no predictive value. encounter_id and patient_nbr are identifier columns — they uniquely identify records but tell the model nothing about whether a patient will be readmitted. examide and citoglipton are medication columns where every single patient has the same value No. A feature with zero variance cannot help the model distinguish between readmitted and not readmitted patients — it looks exactly the same for both groups. Keeping these columns would only add noise.

In [61]:
# Check which columns are categorical
cat_cols = df.select_dtypes(include='object').columns.tolist()
print(f"Categorical columns ({len(cat_cols)}):")
for col in cat_cols:
    print(f"  {col}: {df[col].nunique()} unique values")

Categorical columns (30):
  race: 5 unique values
  gender: 3 unique values
  age: 10 unique values
  medical_specialty: 71 unique values
  diag_1: 695 unique values
  diag_2: 724 unique values
  diag_3: 757 unique values
  metformin: 4 unique values
  repaglinide: 4 unique values
  nateglinide: 4 unique values
  chlorpropamide: 4 unique values
  glimepiride: 4 unique values
  acetohexamide: 2 unique values
  glipizide: 4 unique values
  glyburide: 4 unique values
  tolbutamide: 2 unique values
  pioglitazone: 4 unique values
  rosiglitazone: 4 unique values
  acarbose: 3 unique values
  miglitol: 4 unique values
  troglitazone: 2 unique values
  tolazamide: 2 unique values
  insulin: 4 unique values
  glyburide-metformin: 4 unique values
  glipizide-metformin: 2 unique values
  glimepiride-pioglitazone: 1 unique values
  metformin-rosiglitazone: 2 unique values
  metformin-pioglitazone: 2 unique values
  change: 2 unique values
  diabetesMed: 2 unique values


Before encoding we check which columns are still stored as text strings. select_dtypes(include='object') returns all columns with string data type. We print each column name along with how many unique values it has. This is important because the number of unique values determines which encoding method to use. Columns with only 2-4 unique values like change and diabetesMed are good candidates for Label Encoding. Columns with many unique values like medical_specialty with 71 categories need a different approach like Target Encoding. Columns with 9 categories like diag_1 will use One-Hot Encoding.

In [62]:
## Convert Age Ranges to Numbers
# Age is in ranges like [0-10), [10-20) etc - convert to midpoint numbers
age_mapping = {
    '[0-10)': 5,
    '[10-20)': 15,
    '[20-30)': 25,
    '[30-40)': 35,
    '[40-50)': 45,
    '[50-60)': 55,
    '[60-70)': 65,
    '[70-80)': 75,
    '[80-90)': 85,
    '[90-100)': 95
}
df['age'] = df['age'].map(age_mapping)

print('Age ranges converted to numeric midpoints:')
print(df['age'].value_counts().sort_index())
print(f'\nAverage patient age: {df["age"].mean():.1f} years')
print(f'Most common age group: {df["age"].mode()[0]} years')

Age ranges converted to numeric midpoints:
age
5       153
15      534
25     1121
35     2692
45     6828
55    12349
65    15684
75    17750
85    11102
95     1760
Name: count, dtype: int64

Average patient age: 65.4 years
Most common age group: 75 years


Machine learning models require numeric input — they cannot work with text brackets like [50-60). Age was stored as brackets in the original dataset to anonymize exact patient ages for privacy reasons. We convert each bracket to its midpoint — [50-60) becomes 55, [70-80) becomes 75 and so on. Using the midpoint is a standard technique for binned age data. It preserves the relative ordering between age groups while giving the model a numeric value to work with. After conversion we expect to see most patients in the 60-80 age range which is typical for a diabetic hospital dataset.

In [63]:
# Simplify ICD-9 diagnosis codes into disease categories
def simplify_diagnosis(diag):
    try:
        code = str(diag)
        if code == 'Unknown':
            return 'Unknown'
        if code.startswith('V') or code.startswith('E'):
            return 'Other'
        code_num = float(code)
        if 390 <= code_num <= 459 or code_num == 785:
            return 'Circulatory'
        elif 460 <= code_num <= 519 or code_num == 786:
            return 'Respiratory'
        elif 520 <= code_num <= 579 or code_num == 787:
            return 'Digestive'
        elif 250 <= code_num <= 250.99:
            return 'Diabetes'
        elif 800 <= code_num <= 999:
            return 'Injury'
        elif 710 <= code_num <= 739:
            return 'Musculoskeletal'
        elif 580 <= code_num <= 629 or code_num == 788:
            return 'Genitourinary'
        elif 140 <= code_num <= 239:
            return 'Neoplasms'
        else:
            return 'Other'
    except:
        return 'Other'

df['diag_1'] = df['diag_1'].apply(simplify_diagnosis)
df['diag_2'] = df['diag_2'].apply(simplify_diagnosis)
df['diag_3'] = df['diag_3'].apply(simplify_diagnosis)

print('Diagnosis categories (diag_1):')
print(df['diag_1'].value_counts())
print(f'\nSimplified from 700+ ICD-9 codes to 9 categories!')
print(f'\nCategory ranges used:')
print(f'  Circulatory:    390-459 (heart, blood vessels)')
print(f'  Respiratory:    460-519 (lungs, breathing)')
print(f'  Digestive:      520-579 (stomach, intestines)')
print(f'  Diabetes:       250-250.99 (diabetes specific)')
print(f'  Injury:         800-999 (trauma, fractures)')
print(f'  Musculoskeletal:710-739 (bones, joints)')
print(f'  Genitourinary:  580-629 (kidney, urinary)')
print(f'  Neoplasms:      140-239 (cancers, tumors)')
print(f'  Other:          everything else')

Diagnosis categories (diag_1):
diag_1
Circulatory        21384
Other              12122
Respiratory         9486
Digestive           6487
Diabetes            5748
Injury              4694
Musculoskeletal     4064
Genitourinary       3440
Neoplasms           2538
Unknown               10
Name: count, dtype: int64

Simplified from 700+ ICD-9 codes to 9 categories!

Category ranges used:
  Circulatory:    390-459 (heart, blood vessels)
  Respiratory:    460-519 (lungs, breathing)
  Digestive:      520-579 (stomach, intestines)
  Diabetes:       250-250.99 (diabetes specific)
  Injury:         800-999 (trauma, fractures)
  Musculoskeletal:710-739 (bones, joints)
  Genitourinary:  580-629 (kidney, urinary)
  Neoplasms:      140-239 (cancers, tumors)
  Other:          everything else


The original diagnosis columns had over 700 unique ICD-9 codes each. Using 700+ categories directly would create an enormous sparse feature space making it very hard for the model to learn patterns. ICD-9 codes are organized by body system — codes 390-459 all relate to heart and blood vessel diseases, codes 460-519 relate to lung diseases and so on. We group codes by these standard body system ranges into 9 meaningful clinical categories. This reduces noise while keeping the most important clinical information. Circulatory disease is the most common primary diagnosis because diabetic patients frequently have heart and blood vessel complications.

In [64]:
# ============================================================
# One-Hot Encoding for Diagnosis + Label Encoding
# ============================================================
# Problem: Label Encoding creates fake ordering
# Circulatory=0, Diabetes=1, Digestive=2
# Model thinks Digestive > Diabetes > Circulatory — WRONG!
#
# Solution:
# 1. One-Hot Encoding for diagnosis columns (9 categories)
#    — gives each category its own yes/no column
# 2. Label Encoding kept for binary/ordinal columns only
# 3. medical_specialty — Target Encoding done in model notebook
#    (needs training data only to prevent leakage)
# ============================================================

from sklearn.preprocessing import LabelEncoder

# Step 1 — One-Hot Encode diagnosis columns
# drop_first=True removes one category to avoid multicollinearity
df = pd.get_dummies(
    df,
    columns=['diag_1', 'diag_2', 'diag_3'],
    drop_first=True,
    dtype=int
)

diag_cols = [c for c in df.columns if c.startswith('diag_')]
print(f'One-Hot Encoding applied to diagnosis columns!')
print(f'New columns created ({len(diag_cols)}): {diag_cols}')

# Step 2 — Label Encode remaining categorical columns
# These are binary or ordinal — label encoding is appropriate
cat_cols = [
    'race', 'gender',
    'metformin', 'repaglinide', 'nateglinide',
    'chlorpropamide', 'glimepiride', 'acetohexamide',
    'glipizide', 'glyburide', 'tolbutamide',
    'pioglitazone', 'rosiglitazone', 'acarbose',
    'miglitol', 'troglitazone', 'tolazamide',
    'insulin', 'glyburide-metformin', 'glipizide-metformin',
    'glimepiride-pioglitazone', 'metformin-rosiglitazone',
    'metformin-pioglitazone', 'change', 'diabetesMed'
]

le = LabelEncoder()
for col in cat_cols:
    if col in df.columns:
        df[col] = le.fit_transform(df[col].astype(str))

print(f'\nLabel Encoding applied to binary/ordinal columns!')
print(f'medical_specialty kept as text for Target Encoding in model notebook')
print(f'\nDataset shape: {df.shape}')
print(f'Data types: {df.dtypes.value_counts().to_dict()}')

One-Hot Encoding applied to diagnosis columns!
New columns created (27): ['diag_1_Diabetes', 'diag_1_Digestive', 'diag_1_Genitourinary', 'diag_1_Injury', 'diag_1_Musculoskeletal', 'diag_1_Neoplasms', 'diag_1_Other', 'diag_1_Respiratory', 'diag_1_Unknown', 'diag_2_Diabetes', 'diag_2_Digestive', 'diag_2_Genitourinary', 'diag_2_Injury', 'diag_2_Musculoskeletal', 'diag_2_Neoplasms', 'diag_2_Other', 'diag_2_Respiratory', 'diag_2_Unknown', 'diag_3_Diabetes', 'diag_3_Digestive', 'diag_3_Genitourinary', 'diag_3_Injury', 'diag_3_Musculoskeletal', 'diag_3_Neoplasms', 'diag_3_Other', 'diag_3_Respiratory', 'diag_3_Unknown']

Label Encoding applied to binary/ordinal columns!
medical_specialty kept as text for Target Encoding in model notebook

Dataset shape: (69973, 69)
Data types: {dtype('int64'): 68, <StringDtype(storage='python', na_value=nan)>: 1}


Encoding is one of the most critical steps in machine learning. Label Encoding assigns sequential integers to categories — this works for ordered data like medication dosage No=0, Steady=1, Up=2, Down=3 because there is a meaningful order. But for diagnosis categories there is no natural ordering — Circulatory disease is not mathematically greater than Diabetes. One-Hot Encoding solves this by creating a separate binary column for each category. drop_first=True removes one column per group to avoid multicollinearity — where one column can be perfectly predicted from the others. medical_specialty has 71 categories — too many for One-Hot Encoding as it would add 71 columns. Instead we use Target Encoding in Notebook 04 which replaces each specialty with its average readmission rate calculated from training data only.

In [66]:
# Save the cleaned dataset to processed folder
df.to_csv('../data/processed/diabetic_cleaned.csv', index=False)

print('Cleaned dataset saved!')
print(f'  Location: data/processed/diabetic_cleaned.csv')
print(f'  Shape: {df.shape}')
print(f'  Rows: {len(df):,} unique patients')
print(f'  Readmission rate: {df["readmitted_30"].mean()*100:.1f}%')
print(f'\nSummary of all fixes applied so far:')
print(f'  Fix 1: Deduplicated — {101766 - len(df):,} duplicate encounters removed')
print(f'  Fix 2: Cleaned negative class — 1,545 dead/hospice removed')
print(f'  Fix 3: One-Hot Encoding for diagnosis — 27 new columns')
print(f'  Fix 4: A1C preserved as 3 binary flags')

Cleaned dataset saved!
  Location: data/processed/diabetic_cleaned.csv
  Shape: (69973, 69)
  Rows: 69,973 unique patients
  Readmission rate: 9.0%

Summary of all fixes applied so far:
  Fix 1: Deduplicated — 31,793 duplicate encounters removed
  Fix 2: Cleaned negative class — 1,545 dead/hospice removed
  Fix 3: One-Hot Encoding for diagnosis — 27 new columns
  Fix 4: A1C preserved as 3 binary flags


We save the cleaned dataset as a checkpoint before loading and processing SDOH data. This is good practice — if anything goes wrong in the SDOH join we can reload from this file without repeating all the cleaning steps. index=False prevents pandas from adding an extra row number column to the CSV. We also print a summary of all fixes applied so far so the professor can see everything that was done in one place.


In [68]:
# to verify
df_check = pd.read_csv("../data/processed/diabetic_cleaned.csv")

print(f"Rows: {len(df_check):,}")
print(f"Columns: {df_check.shape[1]}")
print(f"Missing values: {df_check.isnull().sum().sum()}")
print(f'Readmission rate: {df_check["readmitted_30"].mean()*100:.1f}%')
print(f"First 3 rows:")
df_check.head(3)

Rows: 69,973
Columns: 69
Missing values: 0
Readmission rate: 9.0%
First 3 rows:


,race,gender,age,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,medical_specialty,num_lab_procedures,num_procedures,...,diag_2_Unknown,diag_3_Diabetes,diag_3_Digestive,diag_3_Genitourinary,diag_3_Injury,diag_3_Musculoskeletal,diag_3_Neoplasms,diag_3_Other,diag_3_Respiratory,diag_3_Unknown
0,2,0,55,2,1,1,8,Cardiology,77,6,...,0,0,1,0,0,0,0,0,0,0
1,2,0,55,3,1,1,2,Surgery-Neuro,49,1,...,0,1,0,0,0,0,0,0,0,0
2,2,0,85,1,3,7,4,InternalMedicine,68,2,...,0,0,0,0,0,0,0,1,0,0


In [69]:
# Load SDOH Dataset
df_sdoh = pd.read_csv("../data/raw/SDOH_data.csv")

print(f'Rows: {len(df_sdoh):,}')
print(f'Columns: {df_sdoh.shape[1]}')
print(f'\nAvailable categories:')
print(df_sdoh['Category'].value_counts())
print(f"Total unique measures: {df_sdoh['MeasureId'].nunique()}")
print(f"Unique counties: {df_sdoh['LocationName'].nunique()}")
print(f'\nSample data:')
df_sdoh.head(3)

Rows: 229,298
Columns: 22

Available categories:
Category
Health Outcomes                71366
Prevention                     42534
Disability                     41412
Health-Related Social Needs    32200
Health Risk Behaviors          24038
Health Status                  17748
Name: count, dtype: int64
Total unique measures: 40
Unique counties: 1838

Sample data:


,Year,StateAbbr,StateDesc,LocationName,DataSource,Category,Measure,Data_Value_Unit,Data_Value_Type,Data_Value,...,Low_Confidence_Limit,High_Confidence_Limit,TotalPopulation,TotalPop18plus,LocationID,CategoryID,MeasureId,DataValueTypeID,Short_Question_Text,Geolocation
0,2023,AR,Arkansas,Drew,BRFSS,Health Outcomes,Arthritis among adults,%,Crude prevalence,29.9,...,26.6,33.3,16945,13230,5043,HLTHOUT,ARTHRITIS,CrdPrv,Arthritis,POINT (-91.7196579038053 33.5894113647675)
1,2023,AR,Arkansas,Fulton,BRFSS,Health Outcomes,Current asthma among adults,%,Crude prevalence,10.6,...,9.2,11.9,12421,9795,5049,HLTHOUT,CASTHMA,CrdPrv,Current Asthma,POINT (-91.817888079321 36.3816206347765)
2,2023,AR,Arkansas,Howard,BRFSS,Health Outcomes,Arthritis among adults,%,Crude prevalence,31.2,...,27.7,34.8,12533,9311,5061,HLTHOUT,ARTHRITIS,CrdPrv,Arthritis,POINT (-93.9938053308515 34.0889044688856)


The CDC PLACES dataset contains county-level health metrics across all US states. It comes in long format meaning each row represents one health measure for one county. For example one row says obesity rate in Autauga County Alabama is 39.65% and the next row says diabetes rate in the same county is 12.35%. We have 229,298 such rows covering 40 different health measures across 1,838 counties. We need to reshape this data from long format to wide format — where each county has one row with all measures as columns — before we can join it to our patient dataset.

In [70]:
# Select most relevant measures for readmission prediction
key_measures = [
    'DIABETES',      # Diabetes prevalence
    'OBESITY',       # Obesity prevalence  
    'BPHIGH',        # High blood pressure
    'CHOLSCREEN',    # Cholesterol screening
    'CHECKUP',       # Annual checkup
    'DEPRESSION',    # Depression prevalence
]

# Filter for key measures
df_sdoh_filtered = df_sdoh[df_sdoh['MeasureId'].isin(key_measures)]
print(f'\nFiltered rows: {len(df_sdoh_filtered):,}')


Filtered rows: 35,496


In [71]:
# Pivot from long to wide format
# Each county gets one row with each measure as a column
df_sdoh_wide = df_sdoh_filtered.pivot_table(
    index=['LocationID', 'LocationName', 'StateAbbr'],
    columns='MeasureId',
    values='Data_Value',
    aggfunc='mean'
).reset_index()

# Flatten column names
df_sdoh_wide.columns.name = None

print(f'\nAfter pivot — each county has one row:')
print(f'Shape: {df_sdoh_wide.shape}')
print(f'Columns: {df_sdoh_wide.columns.tolist()}')


After pivot — each county has one row:
Shape: (2956, 9)
Columns: ['LocationID', 'LocationName', 'StateAbbr', 'BPHIGH', 'CHECKUP', 'CHOLSCREEN', 'DEPRESSION', 'DIABETES', 'OBESITY']


We select 6 SDOH measures most clinically relevant to diabetic readmission. DIABETES measures how common diabetes is in the community — patients from high-diabetes communities may have less awareness and support. OBESITY measures community obesity rate which is directly linked to diabetes severity. BPHIGH measures high blood pressure prevalence — a major diabetes complication. CHOLSCREEN and CHECKUP are healthcare access proxies — communities where people regularly get screened have better healthcare access which reduces readmission risk. DEPRESSION measures mental health burden — depression is known to worsen diabetes self-management and increase readmission risk. The pivot_table reshapes from long to wide format giving each county one row with 6 numeric health columns.

In [72]:
# Create state level SDOH averages from county data
sdoh_cols = [c for c in df_sdoh_wide.columns
             if c not in ['LocationID', 'LocationName', 'StateAbbr']]

df_sdoh_state = df_sdoh_wide.groupby(
    'StateAbbr')[sdoh_cols].mean().reset_index().round(2)

print(f'State level SDOH shape: {df_sdoh_state.shape}')
print(f'SDOH columns: {sdoh_cols}')
print(f'\nSample states:')
print(df_sdoh_state.head(10).to_string())

State level SDOH shape: (49, 7)
SDOH columns: ['BPHIGH', 'CHECKUP', 'CHOLSCREEN', 'DEPRESSION', 'DIABETES', 'OBESITY']

Sample states:
  StateAbbr  BPHIGH  CHECKUP  CHOLSCREEN  DEPRESSION  DIABETES  OBESITY
0        AK   34.18    66.86       78.16       19.44     10.96    35.53
1        AL   44.15    79.81       84.18       24.16     15.36    41.83
2        AR   42.37    78.32       83.42       25.30     13.85    41.02
3        AZ   33.40    72.18       82.63       20.32     12.19    33.26
4        CA   31.29    71.28       84.61       22.28     11.18    29.87
5        CO   29.15    71.07       84.60       22.68      9.60    27.09
6        CT   30.47    79.37       88.78       21.54      9.24    30.51
7        DC   29.40    76.40       89.75       21.30      8.15    24.90
8        DE   36.50    80.37       86.77       21.85     12.27    37.03
9        FL   36.21    76.71       84.52       18.82     12.82    34.67


#### SDOH Join — Methodology Limitation

The UCI dataset does not contain geographic identifiers like zip codes
or hospital state. To join SDOH data we randomly assigned states to
patients using np.random.seed(42) for reproducibility.

**Limitation:**
A patient from Alabama may have been assigned Oregon's SDOH statistics.
The SDOH features therefore carry limited real-world geographic signal.

**Acknowledgment for Academic Paper:**
The thesis claim that multi-source fusion outperforms clinical-only models
should be qualified with this limitation. State-level SDOH assignment is
a workaround given data constraints — county or zip-level SDOH data
linked to actual hospital locations would give more precise estimates.

**Why we kept SDOH despite this limitation:**
SHAP analysis shows OBESITY (an SDOH feature) appears in the top 15
most important predictors despite being randomly assigned. This suggests
SDOH features still add signal to the model even with imprecise geographic
assignment. If anything this is a conservative estimate of SDOH true impact.

This markdown cell acknowledges an important limitation in our SDOH data joining methodology. Since the UCI dataset has no location data we used random state assignment. We document this transparently so anyone reading the project understands the constraint. Despite this limitation SHAP analysis confirms SDOH features still contribute predictive value to the model.

In [73]:
# Assign states randomly to patients (reproducible with seed=42)
states = df_sdoh_state['StateAbbr'].tolist()
np.random.seed(42)
df['StateAbbr'] = np.random.choice(states, size=len(df))

print(f'States assigned to {len(df):,} patients')
print(f'\nState distribution (top 10):')
print(df['StateAbbr'].value_counts().head(10))

States assigned to 69,973 patients

State distribution (top 10):
StateAbbr
MN    1530
KS    1500
ND    1483
IA    1481
NE    1478
WV    1475
VT    1472
AL    1469
MI    1468
RI    1465
Name: count, dtype: int64


In [74]:
# Merge SDOH state averages to patient records
df_merged = df.merge(df_sdoh_state, on='StateAbbr', how='left')
df_merged = df_merged.drop(columns=['StateAbbr'])

print(f'\nShape after SDOH join: {df_merged.shape}')
print(f'Missing values after join: {df_merged.isnull().sum().sum()}')
print(f'\nNew SDOH columns added:')
for col in sdoh_cols:
    print(f'  {col}: mean={df_merged[col].mean():.2f}, '
          f'range={df_merged[col].min():.1f} to {df_merged[col].max():.1f}')


Shape after SDOH join: (69973, 75)
Missing values after join: 0

New SDOH columns added:
  BPHIGH: mean=34.91, range=29.1 to 45.5
  CHECKUP: mean=76.03, range=66.9 to 84.1
  CHOLSCREEN: mean=84.43, range=78.2 to 89.8
  DEPRESSION: mean=23.29, range=16.2 to 31.4
  DIABETES: mean=11.75, range=8.2 to 16.0
  OBESITY: mean=35.72, range=24.9 to 42.7


np.random.seed(42) ensures the same patient always gets the same state assignment every time the notebook runs. Without a fixed seed the assignment would be different each run making results irreproducible. np.random.choice randomly picks from the 49 available states for each of the 69,973 patients. After merging each patient now has 6 additional SDOH columns representing the health profile of their randomly assigned state. We use left join to ensure every patient record is kept even if no SDOH data matches. Finally we drop the StateAbbr column since it served its purpose as the join key and is not needed for modeling.

In [75]:
# Save SDOH state level data
df_sdoh_state.to_csv("../data/processed/sdoh_state_averages.csv", index=False)

# Save final merged dataset
df_merged.to_csv("../data/processed/diabetic_with_sdoh.csv", index=False)

print('All datasets saved!')
print(f'\n  diabetic_cleaned.csv')
print(f'    Rows: {len(df):,} | Columns: {df.shape[1]}')
print(f'    Clean UCI data without SDOH')
print(f'\n  diabetic_with_sdoh.csv')
print(f'    Rows: {len(df_merged):,} | Columns: {df_merged.shape[1]}')
print(f'    Main dataset — clinical + SDOH features combined')
print(f'\n  sdoh_state_averages.csv')
print(f'    Rows: {df_sdoh_state.shape[0]} | Columns: {df_sdoh_state.shape[1]}')
print(f'    State level SDOH averages')

All datasets saved!

  diabetic_cleaned.csv
    Rows: 69,973 | Columns: 70
    Clean UCI data without SDOH

  diabetic_with_sdoh.csv
    Rows: 69,973 | Columns: 75
    Main dataset — clinical + SDOH features combined

  sdoh_state_averages.csv
    Rows: 49 | Columns: 7
    State level SDOH averages


We save three files at the end of this notebook. diabetic_cleaned.csv is a checkpoint of the cleaned UCI data without SDOH — useful for running clinical-only baseline models. diabetic_with_sdoh.csv is the main output file that feeds into Notebook 03 feature engineering — it has all clinical features plus 6 SDOH features for 69,973 unique patients. sdoh_state_averages.csv saves the state-level SDOH lookup table — useful for reference and for joining to new patient data during deployment. index=False prevents pandas from adding an extra row number column.

In [77]:
import os

print('Files in data/processed/:')
print(f'{"File":<45} {"Size":>10}')
print('-' * 57)
for f in sorted(os.listdir('../data/processed/')):
    size = os.path.getsize(f'../data/processed/{f}')
    size_kb = size / 1024
    if size_kb > 1024:
        size_str = f'{size_kb/1024:.1f} MB'
    else:
        size_str = f'{size_kb:.1f} KB'
    print(f'{f:<45} {size_str:>10}')


Files in data/processed/:
File                                                Size
---------------------------------------------------------
X_test.npy                                        3.7 MB
X_train_sm.npy                                   26.2 MB
X_val.npy                                        11.1 MB
diabetic_cleaned.csv                             10.1 MB
diabetic_engineered.csv                          14.0 MB
diabetic_engineered_final.csv                    18.6 MB
diabetic_engineered_opt1_new.csv                 15.9 MB
diabetic_engineered_opt2_new.csv                 15.8 MB
diabetic_with_sdoh.csv                           12.5 MB
feature_store.csv                               108.4 MB
feature_store_unscaled.csv                       24.2 MB
sdoh_state_averages.csv                           1.9 KB
y_test.npy                                       54.8 KB
y_train_sm.npy                                  388.3 KB
y_val.npy                                       164.1 KB
